### ISSUE TITLE
 Cannot auto failover

### ISSUE THREAD

**Issue by msj12** [openark/orchestrator#822]
Hello Shlomi,I simulated failover based on the Failure detection method, but the failover failed.
Here are some of my configurations

### MySQL:
```
CHANGE` MASTER TO
  MASTER_HOST='10.186.60.62',
  MASTER_USER='repl',
  MASTER_PASSWORD='123',
  MASTER_PORT=3306,
  MASTER_AUTO_POSITION=1,
  MASTER_CONNECT_RETRY=10;
```
`slave_net_timeout = 2`
`log_bin=on`
`log_slave_updates=on`

### orchestrator:
`"RecoverMasterClusterFilters":  ["*" ]`
`"RecoverIntermediateMasterClusterFilters": ["*"]`

no downtime configuration
global recovery is on
Confirmed the last failover（by kill mysqld）

### deployment：
```
superma-60-62：master
superma-60-63：slave
superma-65-6：orchestrator
```
### step：
I used the iptables command to isolate the superma-60-62, I executed the following command on superma-60-62.
`iptables -A INPUT -s 10.186.65.6 -j DROP`
`iptables -A OUTPUT -s 10.186.65.6 -j DROP`
`iptables -A INPUT -s 10.186.60.63 -j DROP`
`iptables -A OUTPUT -s 10.186.60.63 -j DROP`

The failover did not work as expected.
During failover,  nohup continuously print
```
> WARNING: NamedStopwatch.Stop("instance") IsRunning is false
2019-03-01 14:34:10 ERROR dial tcp 10.186.60.62:3306: i/o timeout
2019-03-01 14:34:10 INFO auditType:emergently-read-topology-instance instance:superma-60-62:3306 cluster:superma-60-62:3306 message:FirstTierSlaveFailingToConnectToMaster
2019-03-01 14:34:11 ERROR dial tcp 10.186.60.62:3306: i/o timeout
2019-03-01 14:34:11 INFO auditType:emergently-read-topology-instance instance:superma-60-62:3306 cluster:superma-60-62:3306 message:FirstTierSlaveFailingToConnectToMaster
2019-03-01 14:34:13 ERROR dial tcp 10.186.60.62:3306: i/o timeout
2019-03-01 14:34:13 INFO auditType:emergently-read-topology-instance instance:superma-60-62:3306 cluster:superma-60-62:3306 message:FirstTierSlaveFailingToConnectToMaster
2019-03-01 14:34:14 ERROR dial tcp 10.186.60.62:3306: i/o timeout
2019-03-01 14:34:14 INFO auditType:emergently-read-topology-instance instance:superma-60-62:3306 cluster:superma-60-62:3306 message:FirstTierSlaveFailingToConnectToMaster
2019-03-01 14:34:15 ERROR dial tcp 10.186.60.62:3306: i/o timeout
```
I don't know where the problem is. Can you help me with this question?

**Comment by shlomi-noach** [comment]
pro tip: use three backtips for multi-line code quotes. See how I've edited your original comment.

**Comment by shlomi-noach** [comment]
Can you please run with `--debug` and paste the entire log that relates to detection + recovery?

**Comment by msj12** [comment]
Hi,shlomi-noach. This is the log I got using the below command. 
`grep "2019-03-01" nohup.out > failover.log`

[failover.log](https://github.com/github/orchestrator/files/2924135/failover.log)

**Comment by shlomi-noach** [comment]
Please, I'd like to help you, but can you kindly provide less than 24 hours of log?

**Comment by shlomi-noach** [comment]
I see a lot of `Access Denied` message; for example:
```
2019-03-01 05:25:03 ERROR ExpireAvailableNodes: failed to remove old entries: Error 1045: Access denied for user 'orchestrator'@'127.0.0.1' (using password: YES)

..
2019-03-01 05:27:03 ERROR Error 1045: Access denied for user 'orchestrator'@'127.0.0.1' (using password: YES)
2019-03-01 05:27:03 ERROR Error 1045: Access denied for user 'orchestrator'@'127.0.0.1' (using password: YES)
...
```
Can you look into credentials?

**Comment by shlomi-noach** [comment]
According to:
```
iptables -A INPUT -s 10.186.65.6 -j DROP
iptables -A OUTPUT -s 10.186.65.6 -j DROP
iptables -A INPUT -s 10.186.60.63 -j DROP
iptables -A OUTPUT -s 10.186.60.63 -j DROP
```
I don't know what `10.186.65.6` IP is, I'd need more details please.

But `10.186.60.63` seems like you also isolated the replica, no? In which case `orchestrator` is unable to see any server, and so cannot failover.

**Comment by shlomi-noach** [comment]
Again, I'm trying to read a 24 hour long log, but here's a successful failover:
```
2019-03-01 15:09:02 INFO CommandRun(echo 'Detected DeadMaster on superma-60-62:3306. Affected replicas: 1' >> /tmp/recovery.log,[])
2019-03-01 15:09:02 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-089172111
2019-03-01 15:09:02 INFO CommandRun: 
2019-03-01 15:09:02 INFO CommandRun successful. exit status 0
2019-03-01 15:09:02 INFO topology_recovery: Completed OnFailureDetectionProcesses hook 1 of 1 in 1.936506ms
2019-03-01 15:09:02 INFO topology_recovery: done running OnFailureDetectionProcesses hooks
2019-03-01 15:09:02 INFO executeCheckAndRecoverFunction: proceeding with DeadMaster recovery on superma-60-62:3306; isRecoverable?: true; skipProcesses: false
2019-03-01 15:09:02 INFO topology_recovery: will handle DeadMaster event on superma-60-62:3306
2019-03-01 15:09:02 INFO auditType:recover-dead-master instance:superma-60-62:3306 cluster:superma-60-62:3306 message:problem found; will recover
2019-03-01 15:09:02 INFO topology_recovery: Running 1 PreFailoverProcesses hooks
2019-03-01 15:09:02 INFO topology_recovery: Running PreFailoverProcesses hook 1 of 1: echo 'Will recover from DeadMaster on superma-60-62:3306' >> /tmp/recovery.log
2019-03-01 15:09:02 INFO CommandRun(echo 'Will recover from DeadMaster on superma-60-62:3306' >> /tmp/recovery.log,[])
2019-03-01 15:09:02 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-447184034
2019-03-01 15:09:02 INFO CommandRun: 
2019-03-01 15:09:02 INFO CommandRun successful. exit status 0
2019-03-01 15:09:02 INFO topology_recovery: Completed PreFailoverProcesses hook 1 of 1 in 1.726858ms
2019-03-01 15:09:02 INFO topology_recovery: done running PreFailoverProcesses hooks
2019-03-01 15:09:02 INFO topology_recovery: RecoverDeadMaster: will recover superma-60-62:3306
2019-03-01 15:09:02 INFO topology_recovery: RecoverDeadMaster: masterRecoveryType=MasterRecoveryGTID
2019-03-01 15:09:02 INFO topology_recovery: RecoverDeadMaster: regrouping replicas via GTID
2019-03-01 15:09:02 DEBUG Stopping 1 replicas via StopReplicationNicely
2019-03-01 15:09:02 INFO Stopped slave nicely on superma-60-63:3306, Self:mysql-bin.000033:234, Exec:mysql-bin.000047:234
2019-03-01 15:09:02 INFO Stopped replication on superma-60-63:3306, Self:mysql-bin.000033:234, Exec:mysql-bin.000047:234
2019-03-01 15:09:02 DEBUG - sorted replica: superma-60-63:3306 mysql-bin.000047:234
2019-03-01 15:09:02 DEBUG GetCandidateReplica: candidate: superma-60-63:3306, ahead: 0, equal: 0, late: 0, break: 0
2019-03-01 15:09:02 INFO topology_recovery: RecoverDeadMaster: promotedReplicaIsIdeal(superma-60-63:3306)
2019-03-01 15:09:02 DEBUG RegroupReplicasGTID: working on 0 replicas
2019-03-01 15:09:02 INFO Started replication on superma-60-63:3306
2019-03-01 15:09:03 DEBUG RegroupReplicasGTID: done
2019-03-01 15:09:03 INFO auditType:regroup-replicas-gtid instance:superma-60-62:3306 cluster:superma-60-62:3306 message:regrouped replicas of superma-60-62:3306 via GTID; promoted superma-60-63:3306
2019-03-01 15:09:03 INFO auditType:begin-downtime instance:superma-60-62:3306 cluster:superma-60-62:3306 message:owner: superma-65-6, reason: lost-in-recovery
2019-03-01 15:09:03 INFO topology_recovery: RecoverDeadMaster: 0 postponed functions
2019-03-01 15:09:03 INFO topology_recovery: checking if should replace promoted replica with a better candidate
2019-03-01 15:09:03 INFO topology_recovery: + checking if promoted replica is the ideal candidate
2019-03-01 15:09:03 INFO topology_recovery: + searching for an ideal candidate
2019-03-01 15:09:03 INFO topology_recovery: + checking if promoted replica is an OK candidate
2019-03-01 15:09:03 INFO topology_recovery: + searching for a candidate
2019-03-01 15:09:03 INFO topology_recovery: + searching for a candidate
2019-03-01 15:09:03 INFO topology_recovery: + found no server to promote on top promoted replica
2019-03-01 15:09:03 INFO topology_recovery: replace-promoted-replica-with-candidate: promoted instance superma-60-63:3306 requires no further action
2019-03-01 15:09:03 INFO topology_recovery: promoted replica: superma-60-63:3306
2019-03-01 15:09:03 INFO auditType:recover-dead-master instance:superma-60-62:3306 cluster:superma-60-62:3306 message:promoted replica: superma-60-63:3306
2019-03-01 15:09:03 INFO topology_recovery: RecoverDeadMaster: successfully promoted superma-60-63:3306
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: promoted server coordinates: mysql-bin.000033:234
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: will apply MySQL changes to promoted master
2019-03-01 15:09:03 INFO Will reset replica on superma-60-63:3306
2019-03-01 15:09:03 INFO auditType:begin-maintenance instance:superma-60-63:3306 cluster:superma-60-62:3306 message:maintenanceToken: 85, owner: superma-65-6, reason: reset replica
2019-03-01 15:09:03 INFO Stopped replication on superma-60-63:3306, Self:mysql-bin.000033:234, Exec:mysql-bin.000047:234
2019-03-01 15:09:03 INFO Reset slave superma-60-63:3306
2019-03-01 15:09:03 INFO auditType:reset-slave instance:superma-60-63:3306 cluster:superma-60-62:3306 message:superma-60-63:3306 replication reset
2019-03-01 15:09:03 INFO auditType:end-maintenance instance:superma-60-63:3306 cluster:superma-60-62:3306 message:maintenanceToken: 85
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: applying RESET SLAVE ALL on promoted master: success=true
2019-03-01 15:09:03 INFO instance superma-60-63:3306 read_only: false
2019-03-01 15:09:03 INFO auditType:read-only instance:superma-60-63:3306 cluster:superma-60-62:3306 message:set as false
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: applying read-only=0 on promoted master: success=true
2019-03-01 15:09:03 INFO topology_recovery: Writing KV [mysql/master/superma-60-62:superma-60-63:3306 mysql/master/superma-60-62/hostname:superma-60-63 mysql/master/superma-60-62/port:3306 mysql/master/superma-60-62/ipv4:10.186.60.63 mysql/master/superma-60-62/ipv6:]
2019-03-01 15:09:03 ERROR dial tcp 10.186.60.62:3306: connect: connection refused
2019-03-01 15:09:03 ERROR dial tcp 10.186.60.62:3306: connect: connection refused
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: applying read-only=1 on demoted master: success=false
2019-03-01 15:09:03 INFO topology_recovery: Running 1 PostMasterFailoverProcesses hooks
2019-03-01 15:09:03 INFO topology_recovery: Running PostMasterFailoverProcesses hook 1 of 1: echo 'Recovered from DeadMaster on superma-60-62:3306. Failed: superma-60-62:3306; Promoted: superma-60-63:3306' >> /tmp/recovery.log
2019-03-01 15:09:03 INFO CommandRun(echo 'Recovered from DeadMaster on superma-60-62:3306. Failed: superma-60-62:3306; Promoted: superma-60-63:3306' >> /tmp/recovery.log,[])
2019-03-01 15:09:03 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-120930201
2019-03-01 15:09:03 INFO CommandRun: 
2019-03-01 15:09:03 INFO CommandRun successful. exit status 0
2019-03-01 15:09:03 INFO topology_recovery: Completed PostMasterFailoverProcesses hook 1 of 1 in 1.79457ms
2019-03-01 15:09:03 INFO topology_recovery: done running PostMasterFailoverProcesses hooks
2019-03-01 15:09:03 INFO topology_recovery: - RecoverDeadMaster: updating cluster_alias: superma-60-62:3306 -> superma-60-63:3306
2019-03-01 15:09:03 INFO Topology recovery: {"Id":38,"UID":"1551424142677818722:fea9984e1380d7458941c3f8b388571012ed682a8599a17119d1773531ebf1c0","AnalysisEntry":{"AnalyzedInstanceKey":{"Hostname":"superma-60-62","Port":3306},"AnalyzedInstanceMasterKey":{"Hostname":"","Port":0},"ClusterDetails":{"ClusterName":"superma-60-62:3306","ClusterAlias":"superma-60-62","ClusterDomain":"","CountInstances":0,"HeuristicLag":0,"HasAutomatedMasterRecovery":true,"HasAutomatedIntermediateMasterRecovery":true},"AnalyzedInstanceDataCenter":"","AnalyzedInstancePhysicalEnvironment":"","IsMaster":true,"IsCoMaster":false,"LastCheckValid":false,"LastCheckPartialSuccess":false,"CountReplicas":1,"CountValidReplicas":1,"CountValidReplicatingReplicas":0,"CountReplicasFailingToConnectToMaster":1,"CountDowntimedReplicas":0,"ReplicationDepth":0,"SlaveHosts":[{"Hostname":"superma-60-63","Port":3306}],"IsFailingToConnectToMaster":false,"Analysis":"DeadMaster","Description":"Master cannot be reached by orchestrator and none of its replicas is replicating","StructureAnalysis":null,"IsDowntimed":false,"IsReplicasDowntimed":false,"DowntimeEndTimestamp":"","DowntimeRemainingSeconds":0,"IsBinlogServer":false,"PseudoGTIDImmediateTopology":false,"OracleGTIDImmediateTopology":true,"MariaDBGTIDImmediateTopology":false,"BinlogServerImmediateTopology":false,"CountStatementBasedLoggingReplicas":0,"CountMixedBasedLoggingReplicas":0,"CountRowBasedLoggingReplicas":1,"CountDistinctMajorVersionsLoggingReplicas":1,"CountDelayedReplicas":0,"CountLaggingReplicas":0,"IsActionableRecovery":true,"ProcessingNodeHostname":"superma-65-6","ProcessingNodeToken":"2739914f3651eedfbf3be23990dbcb05585f90a1fe32c51ed67a923c42ac7782","CountAdditionalAgreeingNodes":0,"StartActivePeriod":"","SkippableDueToDowntime":false,"GTIDMode":"ON","MinReplicaGTIDMode":"ON","MaxReplicaGTIDMode":"ON","MaxReplicaGTIDErrant":"","CommandHint":""},"SuccessorKey":{"Hostname":"superma-60-63","Port":3306},"SuccessorAlias":"","IsActive":false,"IsSuccessful":true,"LostReplicas":[],"ParticipatingInstanceKeys":[],"AllErrors":[],"RecoveryStartTimestamp":"","RecoveryEndTimestamp":"","ProcessingNodeHostname":"","ProcessingNodeToken":"","Acknowledged":false,"AcknowledgedAt":"","AcknowledgedBy":"","AcknowledgedComment":"","LastDetectionId":0,"RelatedRecoveryId":0,"Type":"MasterRecovery","RecoveryType":"MasterRecoveryGTID"}
2019-03-01 15:09:03 INFO topology_recovery: Running 1 PostFailoverProcesses hooks
2019-03-01 15:09:03 INFO topology_recovery: Running PostFailoverProcesses hook 1 of 1: echo '(for all types) Recovered from DeadMaster on superma-60-62:3306. Failed: superma-60-62:3306; Successor: superma-60-63:3306' >> /tmp/recovery.log
2019-03-01 15:09:03 INFO CommandRun(echo '(for all types) Recovered from DeadMaster on superma-60-62:3306. Failed: superma-60-62:3306; Successor: superma-60-63:3306' >> /tmp/recovery.log,[])
2019-03-01 15:09:03 INFO CommandRun/running: bash /tmp/orchestrator-process-cmd-793463332
2019-03-01 15:09:03 INFO CommandRun: 
2019-03-01 15:09:03 INFO CommandRun successful. exit status 0
2019-03-01 15:09:03 INFO topology_recovery: Completed PostFailoverProcesses hook 1 of 1 in 1.584334ms
2019-03-01 15:09:03 INFO topology_recovery: done running PostFailoverProcesses hooks
2019-03-01 15:09:03 INFO topology_recovery: Waiting for 0 postponed functions
2019-03-01 15:09:03 DEBUG PostponedFunctionsContainer: waiting on 0 postponed functions
2019-03-01 15:09:03 DEBUG PostponedFunctionsContainer: done waiting
2019-03-01 15:09:03 INFO topology_recovery: Executed 0 postponed functions
```

I'd really appreciate it if you pinpointed when you ran the network isolation, on which servers, and give me the logs from the relevant time.

**Comment by msj12** [comment]
I am very sorry, it is my fault.
